crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -499,6 +524,103 @@ impl<'a> TimelineEventHandler<'a> {
         self.state.reactions.map.insert(reaction_id, (reaction_sender_data, c.relates_to));
     }
 
+    #[instrument(skip_all, fields(replacement_event_id = ?replacement.event_id))]
+    fn handle_poll_start_edit(
+        &mut self,
+        replacement: Replacement<UnstablePollStartEventContentWithoutRelation>,
+    ) {
+        update_timeline_item!(self, &replacement.event_id, "poll edit", |event_item| {
+            if self.ctx.sender != event_item.sender() {
+                info!(
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "Edit event applies to another user's timeline item, discarding"
+                );
+                return None;
+            }
+
+            let poll_state = if let TimelineItemContent::Poll(poll_state) = &event_item.content() {
+                poll_state
+            } else {
+                info!(
+                    original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                    "Can't edit a poll that is not of type TimelineItemContent::Poll, discarding"
+                );
+                return None;
+            };
+
+            let new_content =
+                if let Ok(edited_poll_state) = poll_state.edit(&replacement.new_content) {
+                    TimelineItemContent::Poll(edited_poll_state)
+                } else {
+                    info!(
+                        original_sender = ?event_item.sender(), edit_sender = ?self.ctx.sender,
+                        "This poll can't be edited, discarding"
+                    );
+                    return None;
+                };
+
+            let edit_json = match &self.ctx.flow {
+                Flow::Local { .. } => None,
+                Flow::Remote { raw_event, .. } => Some(raw_event.clone()),
+            };
+
+            trace!("Applying edit");
+            Some(event_item.with_content(new_content, edit_json))
+        });
+    }
+
+    fn handle_poll_start(&mut self, c: UnstablePollStartEventContent, should_add: bool) {
+        let mut poll_state = PollState::new(c);
+        // TODO: If the flow is not remote we're not able to apply the pending events. Is this a problem?
```

Comment:
Can you remove this TODO and instead make a new GitHub issue after this is merged, about being able to vote on your own poll while it is still in local-echo state? It would be nice to support eventually, but editing a local echo is likewise not yet supported, and it seems like a use case that's not terribly important.

Reply:
I don't think this is the case.
This is for the (perhaps even impossible) case where:
- responses / end events arrive before the start event, so they are cached.
- the start event eventually arrives but it is a local echo. Therefore we're not able to apply the cache to it.

Reply:
That is indeed not possible.

Reply:
Good then, will remove the comment. I couldn't understand why this is not possible though :) Can you offer some background here?

Reply:
I've actually added an `else` with an `info!()` just in case.

Reply:
Local echoes are messages sent by the local user that have not been echoed back by the server yet. When a new one is added, that currently always means it was *just* created so there can't be any other events referencing it yet.

The only other case we will add in the future is restoring local echoes from disk, where I guess there will be a very slim chance that you closed the app before receiving the remote echo for the poll, but got some responses from the server before re-opening the timeline and loading it from disk. In that case, it doesn't really make sense to apply responses to the local echo since we know the full event (remote echo) exists further up in the timeline (so we could use this information to remove the local echo).